### PR TITLE
Do not continue retrying to connect to ZK if the query is killed or over limits

### LINKED ISSUE #47983: ZooKeeperRetriesControl does not respect max_execution_time
```sql
production-02 :) set max_execution_time=5

SET max_execution_time = 5

Query id: 4a859eed-1e59-4d1d-bb8d-0288aff3b2b8

Ok.

0 rows in set. Elapsed: 0.002 sec. 

production-02 :) insert into bi_connector_log(start_datetime) Select now();

INSERT INTO bi_connector_log (start_datetime) SELECT now()

Query id: c77ddf83-05b3-4984-adc5-ea934ecce754

↑ Progress: 1.00 rows, 1.00 B (0.01 rows/s., 0.01 B/s.)                                                 
0 rows in set. Elapsed: 142.639 sec. 

Received exception from server (version 23.3.1):
Code: 159. DB::Exception: Received from clickhouse-02:59000. DB::Exception: Timeout exceeded: elapsed 142.722363052 seconds, maximum: 5. (TIMEOUT_EXCEEDED)

```

I would expect to not be any retries after the 5 seconds are reached.

Logs:
```
2023.03.24 15:59:37.157185 [ 365684 ] {c77ddf83-05b3-4984-adc5-ea934ecce754} <Trace> default.bi_connector_log (73d8dba3-16e1-4abc-9d7c-a02573d52d63) (Replicated OutputStream): ZooKeeperRetriesControl: ReplicatedMergeTreeSink::consume/commitPart: setUserError: error=242 message=Table is in readonly mode: replica_path=/clickhouse/tables/01-01/default.bi_connector_log/replicas/clickhouse-02
2023.03.24 15:59:37.157307 [ 365684 ] {c77ddf83-05b3-4984-adc5-ea934ecce754} <Debug> default.bi_connector_log (73d8dba3-16e1-4abc-9d7c-a02573d52d63) (Replicated OutputStream): ZooKeeperRetriesControl: ReplicatedMergeTreeSink::consume/commitPart: retry limit is reached: retry_count=20 timeout=10000ms error=242 message=Table is in readonly mode: replica_path=/clickhouse/tables/01-01/default.bi_connector_log/replicas/clickhouse-02
2023.03.24 15:59:37.202078 [ 365439 ] {c77ddf83-05b3-4984-adc5-ea934ecce754} <Error> executeQuery: Code: 159. DB::Exception: Timeout exceeded: elapsed 142.722363052 seconds, maximum: 5. (TIMEOUT_EXCEEDED) (version 23.3.1.2537) (from 192.168.1.200:43762) (in query: insert into bi_connector_log(start_datetime) Select now();), Stack trace (when copying this message, always include the lines below):

0. ./build_tsan/./contrib/llvm-project/libcxx/include/exception:0: Poco::Exception::Exception(String const&, int) @ 0x2419611e in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
1. ./build_tsan/./src/Common/Exception.cpp:91: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x14cebf25 in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
2. ./build_tsan/./contrib/llvm-project/libcxx/include/string:1499: DB::Exception::Exception<double, double>(int, FormatStringHelperImpl<std::type_identity<double>::type, std::type_identity<double>::type>, double&&, double&&) @ 0x1d1fb912 in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
3. ./build_tsan/./src/QueryPipeline/ExecutionSpeedLimits.cpp:110: DB::ExecutionSpeedLimits::checkTimeLimit(Stopwatch const&, DB::OverflowMode) const @ 0x1d1fb3ed in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
4. ./build_tsan/./src/Interpreters/ProcessList.cpp:0: DB::QueryStatus::checkTimeLimit() @ 0x1e83ee5d in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
5. ./build_tsan/./src/Processors/Executors/PipelineExecutor.cpp:185: DB::PipelineExecutor::finalizeExecution() @ 0x2003e6b2 in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
6. ./build_tsan/./src/Processors/Executors/PipelineExecutor.cpp:118: DB::PipelineExecutor::execute(unsigned long) @ 0x2003e399 in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
7. ./build_tsan/./src/Processors/Executors/CompletedPipelineExecutor.cpp:56: void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::CompletedPipelineExecutor::execute()::$_0>(DB::CompletedPipelineExecutor::execute()::$_0&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x2003cff3 in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
8. ./build_tsan/./base/base/../base/wide_integer_impl.h:786: ThreadPoolImpl<std::thread>::worker(std::__list_iterator<std::thread, void*>) @ 0x14e0a217 in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
9. ./build_tsan/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:302: void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, long, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x14e0ff62 in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
10. __tsan_thread_start_func @ 0xc7e1da3 in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
11. ? @ 0x7f0ea609abb5 in ?
12. ? @ 0x7f0ea611cd90 in ?

2023.03.24 15:59:37.203515 [ 365439 ] {c77ddf83-05b3-4984-adc5-ea934ecce754} <Trace> default.bi_connector_log (73d8dba3-16e1-4abc-9d7c-a02573d52d63): Part all_2_2_0 in temporary state can be removed without unlocking shared state
2023.03.24 15:59:37.204477 [ 365439 ] {c77ddf83-05b3-4984-adc5-ea934ecce754} <Error> TCPHandler: Code: 159. DB::Exception: Timeout exceeded: elapsed 142.722363052 seconds, maximum: 5. (TIMEOUT_EXCEEDED), Stack trace (when copying this message, always include the lines below):

0. ./build_tsan/./contrib/llvm-project/libcxx/include/exception:0: Poco::Exception::Exception(String const&, int) @ 0x2419611e in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
1. ./build_tsan/./src/Common/Exception.cpp:91: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x14cebf25 in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
2. ./build_tsan/./contrib/llvm-project/libcxx/include/string:1499: DB::Exception::Exception<double, double>(int, FormatStringHelperImpl<std::type_identity<double>::type, std::type_identity<double>::type>, double&&, double&&) @ 0x1d1fb912 in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
3. ./build_tsan/./src/QueryPipeline/ExecutionSpeedLimits.cpp:110: DB::ExecutionSpeedLimits::checkTimeLimit(Stopwatch const&, DB::OverflowMode) const @ 0x1d1fb3ed in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
4. ./build_tsan/./src/Interpreters/ProcessList.cpp:0: DB::QueryStatus::checkTimeLimit() @ 0x1e83ee5d in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
5. ./build_tsan/./src/Processors/Executors/PipelineExecutor.cpp:185: DB::PipelineExecutor::finalizeExecution() @ 0x2003e6b2 in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
6. ./build_tsan/./src/Processors/Executors/PipelineExecutor.cpp:118: DB::PipelineExecutor::execute(unsigned long) @ 0x2003e399 in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
7. ./build_tsan/./src/Processors/Executors/CompletedPipelineExecutor.cpp:56: void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::CompletedPipelineExecutor::execute()::$_0>(DB::CompletedPipelineExecutor::execute()::$_0&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x2003cff3 in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
8. ./build_tsan/./base/base/../base/wide_integer_impl.h:786: ThreadPoolImpl<std::thread>::worker(std::__list_iterator<std::thread, void*>) @ 0x14e0a217 in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
9. ./build_tsan/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:302: void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, long, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x14e0ff62 in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
10. __tsan_thread_start_func @ 0xc7e1da3 in /mnt/ch/ClickHouse/build_tsan/programs/clickhouse
11. ? @ 0x7f0ea609abb5 in ?
12. ? @ 0x7f0ea611cd90 in ?

2023.03.24 15:59:37.204633 [ 365439 ] {c77ddf83-05b3-4984-adc5-ea934ecce754} <Debug> TCPHandler: Processed in 142.769807919 sec.
```

It seems really easy to handle by passing a ProcessListElement (`context->getProcessListElement()`) and checking if the query has been killed or timeout (checkTimeLimitSoft + checkTimeLimit) before attempting any retry.

### PR DESCRIPTION
**Pull request by Algunenano**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Do not continue retrying to connect to ZK if the query is killed or over limits

### Documentation entry for user-facing changes

Closes https://github.com/ClickHouse/ClickHouse/issues/47983

Check with a restarting ZK (keeps resetting the session)
```
set max_execution_time=5
insert into bi_connector_log(start_datetime) Select now();
```

* Before: `Code: 159. DB::Exception: Received from clickhouse-02:59000. DB::Exception: Timeout exceeded: elapsed 142.722363052 seconds, maximum: 5. (TIMEOUT_EXCEEDED)`
* After: `Code: 159. DB::Exception: Received from clickhouse-02:59000. DB::Exception: Timeout exceeded: elapsed 7.13354754 seconds, maximum: 5. (TIMEOUT_EXCEEDED)`

### CODE REVIEW

**Comment by Algunenano** [comment]
Added an integration test and rebased with master to fix broken test (timezone issues)

**Comment by Algunenano** [comment]
* `Integration tests (tsan) [4/6] ` -> Likely my fault. I assumed the ZK status/connection was reset per test but probably that's not the case.
* `Stateless tests (tsan) [1/5] ` -> https://github.com/ClickHouse/ClickHouse/issues/47460

File: tests/integration/test_inserts_with_keeper_retries/test.py [hunk 1]
```diff
@@ -98,3 +98,19 @@ def keeper_disconnect(node, event):
...
+
+    cluster.stop_zookeeper_nodes(["zoo1", "zoo2", "zoo3"])
+
+    start_time = time.time()
+    with pytest.raises(QueryRuntimeException):
+        node1.query(
+            "INSERT INTO zk_down SELECT number, toString(number) FROM numbers(10) SETTINGS insert_keeper_max_retries=10000, insert_keeper_retry_max_backoff_ms=1000, max_execution_time=1"
+        )
+    finish_time = time.time()
+    assert finish_time - start_time < 10
```
Comment (SmitaRKulkarni) [inline tests/integration/test_inserts_with_keeper_retries/test.py:116]: looks like you need to use `start_zookeeper_nodes`

**Review by SmitaRKulkarni — COMMENTED** [review]
Rest all LGTM

**Review by SmitaRKulkarni — APPROVED** [review]

**Review by SmitaRKulkarni — APPROVED** [review]

**Comment by SmitaRKulkarni** [comment]
@Algunenano  Could you please resolve conflicts with master  (src/Backups/BackupCoordinationRemote.cpp)

**Comment by SmitaRKulkarni** [comment]
Unrelated failure
Stateless tests (tsan) - 02187_async_inserts_all_formats - #47866

**Review by SmitaRKulkarni — APPROVED** [review]

**Pull request merged**